Simulation segfaults on Node error

When an error occurs in the Mir Node, the simulation might segfault.
A more graceful stop would be nice.
I discovered this when running a test and accidentally misconfigured the modules, such that the destination module of some events was an empty string. Here's the log:

```
>>>>>>>>>>>  libp2p 10 requests and 4 nodes
Node 0: running
Node 1: running
Node 2: running
Node 3: running
Connecting to node: 127.0.0.1:20000
Connecting to node: 127.0.0.1:20001
Connecting to node: 127.0.0.1:20003
Connecting to node: 127.0.0.1:20002
Node 0: ReqRec: Listening for request connections on port 20000
Node 1: ReqRec: Listening for request connections on port 20001
Node 2: ReqRec: Listening for request connections on port 20002
node 1 handler starting on [/ip4/127.0.0.1/tcp/10001]
Node 3: ReqRec: Listening for request connections on port 20003
node 2 handler starting on [/ip4/127.0.0.1/tcp/10002]
node 1 is connecting to node 3
node 2 is connecting to node 0
node 1 is connecting to node 0
node 2 is connecting to node 1
node 1 is connecting to node 2
node 2 is connecting to node 3
node 3 handler starting on [/ip4/127.0.0.1/tcp/10003]
node 3 is connecting to node 0
node 3 is connecting to node 2
node 3 is connecting to node 1
node 0 handler starting on [/ip4/127.0.0.1/tcp/10000]
node 0 is connecting to node 1
node 0 is connecting to node 2
node 0 is connecting to node 3
Node connected. id=2 addr=127.0.0.1:20002
Node 2: ReqRec: Incoming connection from 127.0.0.1:53388
Node connected. id=3 addr=127.0.0.1:20003
Node connected. id=0 addr=127.0.0.1:20000
Node 3: ReqRec: Incoming connection from 127.0.0.1:53390
Node 0: ReqRec: Incoming connection from 127.0.0.1:53391
Node connected. id=1 addr=127.0.0.1:20001
Node 3: ReqRec: Received request clId=1 reqNo=0
Node 0: ReqRec: Received request clId=1 reqNo=0
Node 2: ReqRec: Received request clId=1 reqNo=0
Node 1: ReqRec: Incoming connection from 127.0.0.1:53389
Node 1: ReqRec: Received request clId=1 reqNo=0
node 2 has connected to node 0
node 3 has connected to node 1
node 1 has connected to node 2
node 2 has connected to node 1
node 1 has connected to node 3
node 1 has connected to node 0
node 2 has connected to node 3
node 0 has connected to node 1
Node 1: ReqRec: Received request clId=1 reqNo=1
node 3 has connected to node 2
Node 2: ReqRec: Received request clId=1 reqNo=1
Node 1: ReqRec: Received request clId=1 reqNo=2
Node 1: ReqRec: Received request clId=1 reqNo=3
Node 2: ReqRec: Received request clId=1 reqNo=2
ISS: PBFT: Initializing new view. epoch=0 instance=0 view=0
Node 1: ReqRec: Received request clId=1 reqNo=4
Node 1: ReqRec: Received request clId=1 reqNo=5
ISS: PBFT: Initializing new view. epoch=0 instance=1 view=0
ISS: PBFT: Initializing new view. epoch=0 instance=2 view=0
ISS: PBFT: Initializing new view. epoch=0 instance=3 view=0
Node 2: ReqRec: Received request clId=1 reqNo=3
ISS: PBFT: Initializing new view. epoch=0 instance=0 view=0
ISS: PBFT: Initializing new view. epoch=0 instance=1 view=0
ISS: PBFT: Initializing new view. epoch=0 instance=2 view=0
ISS: PBFT: Initializing new view. epoch=0 instance=3 view=0
Node 1: ReqRec: Received request clId=1 reqNo=6
Node 1: ReqRec: Could not submit request (1-6): no buffer for module  (adding event of type *eventpb.Event_HashResult). Closing connection.
Node 2: ReqRec: Received request clId=1 reqNo=4
Node 2: ReqRec: Received request clId=1 reqNo=5
Node 2: ReqRec: Received request clId=1 reqNo=6
Node 2: ReqRec: Received request clId=1 reqNo=7
Node 2: ReqRec: Received request clId=1 reqNo=8
node 3 has connected to node 0
Node 2: ReqRec: Received request clId=1 reqNo=9
Node 2: ReqRec: Connection terminated: 127.0.0.1:53388 (EOF)
node 0 has connected to node 2
Node 3: ReqRec: Received request clId=1 reqNo=1
node 0 has connected to node 3
Node 3: ReqRec: Received request clId=1 reqNo=2
Node 3: ReqRec: Received request clId=1 reqNo=3
ISS: PBFT: Initializing new view. epoch=0 instance=0 view=0
Node 3: ReqRec: Received request clId=1 reqNo=4
ISS: PBFT: Initializing new view. epoch=0 instance=1 view=0
Node 3: ReqRec: Received request clId=1 reqNo=5
Node 3: ReqRec: Received request clId=1 reqNo=6
ISS: PBFT: Initializing new view. epoch=0 instance=2 view=0
ISS: PBFT: Initializing new view. epoch=0 instance=3 view=0
Node 0: ReqRec: Received request clId=1 reqNo=1
Node 3: ReqRec: Received request clId=1 reqNo=7
ISS: PBFT: Initializing new view. epoch=0 instance=0 view=0
ISS: PBFT: Initializing new view. epoch=0 instance=1 view=0
ISS: PBFT: Initializing new view. epoch=0 instance=2 view=0
ISS: PBFT: Initializing new view. epoch=0 instance=3 view=0
Node 0: ReqRec: Received request clId=1 reqNo=2
Node 3: ReqRec: Received request clId=1 reqNo=8
Node 3: ReqRec: Could not submit request (1-8): no buffer for module  (adding event of type *eventpb.Event_HashResult). Closing connection.
Node 0: ReqRec: Received request clId=1 reqNo=3
Node 0: ReqRec: Received request clId=1 reqNo=4
Node 0: ReqRec: Received request clId=1 reqNo=5
Node 0: ReqRec: Could not submit request (1-5): no buffer for module  (adding event of type *eventpb.Event_HashResult). Closing connection.
Node 2: Node run returned!
Node 2: ReqRec: Stopping request receiver.
Node 1: Node run returned!
Node 1: ReqRec: Stopping request receiver.
Node 2: ReqRec: Request receiver stopped.
Node 2: Fake request submission done.
Stopping libp2p transport.
Closing connection to=0
Node 1: ReqRec: Request receiver stopped.
Node 1: Fake request submission done.
Stopping libp2p transport.
Closing connection to=2
Closed connection to=2
Closing connection to=3
Closed connection to=0
Closing connection to=1
Closed connection to=3
Closing connection to=0
Closed connection to=1
Closing connection to=3
Closed connection to=0
mir handler for 12D3KooWHH-3-11 started
Closed connection to=3
failed to read mir transport request err=EOF
mir handler closing stream for 12D3KooWHH-3-11
mir handler for 12D3KooWHH-2-8 started
mir handler for 12D3KooWHH-3-11 stopped
mir handler for 12D3KooWKR-4-12 started
failed to read mir transport request err=EOF
mir handler for 12D3KooWHH-5-13 started
failed to read mir transport request err=EOF
mir handler closing stream for 12D3KooWHH-5-13
mir handler for 12D3KooWHH-5-13 stopped
mir handler for 12D3KooWKR-1-7 started
failed to read mir transport request err=EOF
mir handler closing stream for 12D3KooWKR-1-7
failed to read mir transport request err=EOF
mir handler closing stream for 12D3KooWKR-4-12
mir handler for 12D3KooWKR-4-12 stopped
mir handler closing stream for 12D3KooWHH-2-8
mir handler for 12D3KooWHH-2-8 stopped
mir handler for 12D3KooWKR-1-7 stopped
node 2 libp2p host closed
Stopping libp2p transport finished.
node 1 libp2p host closed
Stopping libp2p transport finished.
Node 0: Node run returned!
Node 0: ReqRec: Stopping request receiver.
Intercepted events written to event log: 10
Node 3: Node run returned!Intercepted events written to event log: 12

Node 3: ReqRec: Stopping request receiver.
Node 1: Interceptor: Interceptor returned with error. err=interceptor stopped at caller request
Node 0: ReqRec: Request receiver stopped.
Node 0: Fake request submission done.
Stopping libp2p transport.
Closing connection to=1
Node 3: ReqRec: Request receiver stopped.
Node 3: Fake request submission done.
Stopping libp2p transport.
Closing connection to=1
Node 2: Interceptor: Interceptor returned with error. err=interceptor stopped at caller request
Could not close connection to node 1: stream reset
Closing connection to=2
Could not close connection to node 1: stream reset
Closing connection to=2
Could not close connection to node 2: stream reset
Closing connection to=3
Could not close connection to node 2: stream reset
Closing connection to=0
Closed connection to=3
Closed connection to=0
node 3 libp2p host closed
Stopping libp2p transport finished.
node 0 libp2p host closed
Stopping libp2p transport finished.
Node 1: exit with error err=no buffer for module  (adding event of type *eventpb.Event_HashResult)
Node 2: exit with errorIntercepted events written to event log: 12
Intercepted events written to event log: 11
 err=no buffer for module  (adding event of type *eventpb.Event_HashResult)
Node 0: Interceptor: Interceptor returned with error. err=interceptor stopped at caller request
Node 3: Interceptor: Interceptor returned with error. err=interceptor stopped at caller request
Node 0: exit with error err=no buffer for module  (adding event of type *eventpb.Event_HashResult)
Node 3: exit with error err=no buffer for module  (adding event of type *eventpb.Event_HashResult)
All go routines shut down
ISS: New epoch epochNr=0
>>>>>>>>>>>  
Node 0: running
Node 0: ReqRec: Listening for request connections on port 20000
ISS: PBFT: Initializing new view. epoch=0 instance=0 view=0
ISS: PBFT: Proposing. epoch=0 instance=0 sn=0 batchSize=0
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x1008a4130]

goroutine 5783 [running]:
github.com/filecoin-project/mir/pkg/testsim.(*Chan).send(0x0, 0xc0007582c0, {0x1015697c0, 0xc000422410})
	/Users/mato/vmshare/parallels-vm/github.com/matejpavlovic/mir/pkg/testsim/chan.go:32 +0x40
github.com/filecoin-project/mir/pkg/testsim.(*Process).Send(...)
	/Users/mato/vmshare/parallels-vm/github.com/matejpavlovic/mir/pkg/testsim/process.go:67
github.com/filecoin-project/mir/pkg/deploytest.(*SimNode).SendEvents(0xc000778b10, 0x0?, 0xc00099e4c8)
	/Users/mato/vmshare/parallels-vm/github.com/matejpavlovic/mir/pkg/deploytest/simulation.go:84 +0x620
github.com/filecoin-project/mir/pkg/deploytest.(*simModule).run.func1(0x0?)
	/Users/mato/vmshare/parallels-vm/github.com/matejpavlovic/mir/pkg/deploytest/simulation.go:238 +0xa4
created by github.com/filecoin-project/mir/pkg/deploytest.(*simModule).run
	/Users/mato/vmshare/parallels-vm/github.com/matejpavlovic/mir/pkg/deploytest/simulation.go:236 +0x50
FAIL	github.com/filecoin-project/mir/pkg/iss	4.460s
```

--- a/pkg/deploytest/simtransport.go
+++ b/pkg/deploytest/simtransport.go
@@ -147,7 +147,10 @@ func (m *simTransportModule) sendMessage(ctx context.Context, msg *messagepb.Mes
 			return
 		}
 
-		destModule := m.SimTransport.nodes[target]
+		destModule, ok := m.SimTransport.nodes[target]
+		if !ok {
+			panic(fmt.Sprintf("Destination node does not exist: %v", target))
+		}
 		proc.Send(destModule.simChan, message{m.id, target, msg})
 
 		proc.Exit()

--- a/pkg/deploytest/simulation.go
+++ b/pkg/deploytest/simulation.go
@@ -6,6 +6,7 @@ package deploytest
 
 import (
 	"context"
+	"fmt"
 	"math/rand"
 	"time"
 
@@ -81,7 +82,11 @@ func (n *SimNode) SendEvents(proc *testsim.Process, eventList *events.EventList)
 	})
 
 	for _, m := range moduleIDs {
-		proc.Send(n.moduleChans[m.Top()], eventsMap[m.Top()])
+		ch, ok := n.moduleChans[m.Top()]
+		if !ok {
+			panic(fmt.Sprintf("destination module does not exist: %v", m.Top()))
+		}
+		proc.Send(ch, eventsMap[m.Top()])
 		proc.Yield() // wait until the receiver blocks
 	}
 }

